rust/src/operations/create.rs

```diff
@@ -1,251 +1,367 @@
+        // We set the lowest protocol we can, and if subsequent writes use newer features we update metadata?
+        let protocol = self
+            .actions
+            .iter()
+            .find(|a| matches!(a, Action::protocol(_)))
+            .map(|a| match a {
+                Action::protocol(p) => p.clone(),
+                _ => unreachable!(),
+            })
+            .unwrap_or_else(|| Protocol {
+                min_reader_version: MAX_SUPPORTED_READER_VERSION,
+                min_writer_version: MAX_SUPPORTED_WRITER_VERSION,
+            });
+
+        let metadata = DeltaTableMetaData::new(
+            self.name,
+            self.comment,
+            None,
+            SchemaTypeStruct::new(self.columns),
+            self.partition_columns.unwrap_or_default(),
+            HashMap::new(),
+        );
+
+        let operation = DeltaOperation::Create {
+            mode: self.mode.clone(),
+            metadata: metadata.clone(),
+            location: storage_url.to_string(),
+            protocol: protocol.clone(),
+        };
+
+        let mut actions = vec![
+            Action::protocol(protocol),
+            Action::metaData(MetaData::try_from(metadata)?),
+        ];
+        actions.extend(
+            self.actions
+                .into_iter()
+                .filter(|a| matches!(a, Action::protocol(_))),
+        );
+
+        Ok((table, actions, operation))
     }
-    .map_err(to_datafusion_err)?;
+}
+
+impl std::future::IntoFuture for CreateBuilder {
+    type Output = DeltaResult<DeltaTable>;
+    type IntoFuture = BoxFuture<'static, Self::Output>;
 
-    let serialized_batch = serialize_actions(actions)?;
-    let metrics = ExecutionPlanMetricsSet::new();
-    let tracking_metrics = MemTrackingMetrics::new(&metrics, partition_id);
-    let stream = SizedRecordBatchStream::new(
-        serialized_batch.schema(),
-        vec![Arc::new(serialized_batch)],
-        tracking_metrics,
-    );
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            let mode = this.mode.clone();
+            let (mut table, actions, operation) = this.into_table_and_actions()?;
+            if table.object_store().is_delta_table_location().await? {
+                match mode {
+                    SaveMode::ErrorIfExists => return Err(CreateError::TableAlreadyExists.into()),
+                    SaveMode::Append => return Err(CreateError::AppendNotAllowed.into()),
+                    SaveMode::Ignore => {
+                        table.load().await?;
+                        return Ok(table);
+                    }
+                    SaveMode::Overwrite => {
+                        let curr_files =
+                            flatten_list_stream(table.object_store().as_ref(), None).await?;
+                        table.object_store().delete_batch(&curr_files).await?;
```

Comment:
Why delete the current files?

Reply:
Well .. in this case I was not sure, but conceptually I felt we are not overwriting the data in a table and creating a new version, but creating an entirely new table at version 0. If we were to support also updating table metadata, schema etc via this route, I guess there is more work to be done here validating all changes.

But this one I was definitely unsure about.

Reply:
Had a look at the spark implementation - seems they agree with you and updating the metadata and evoling the table is the way to go

https://github.com/delta-io/delta/blob/1f6ab824e14794c17202b5e4e5df6a95357a799c/core/src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala#L196-L208

Reply:
Since the metadata update is a larger operation, I opted for raising not implemented for now in this PR. I opened #917 to track this.